File: Data Structures & Algorithms/jump-game/submission-0.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        // Always start at first one
        // DP + top down recursion?
        unordered_map<int,bool> memo;
        return dfs(nums,0,memo);
    }
private:
    bool dfs(vector<int>& nums, int i, unordered_map<int,bool>& memo) {
        if (memo.count(i)) return memo[i];
        if (i+nums[i]>=nums.size()-1) return true;

        for(int j=1; j<=nums[i]; j++) {
            if (dfs(nums, i+j, memo)){
                memo[i+j] = true;
                return true;
            }
        }
        memo[i] = false;
        return false;
    }
};

File: Data Structures & Algorithms/palindromic-substrings/submission-0.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        memo = {}
        res = 0
        def dp(l:int, r:int) -> int:
            if memo.get((l,r)):
                return memo[(l,r)]
            if l<0 or r >= len(s):
                memo[(l,r)] = 0
                return memo[(l,r)]
            if s[l] != s[r]:
                memo[(l,r)] = 0
                return memo[(l,r)]
            else:
                memo[(l,r)] = 1 + dp(l-1,r+1)
                return memo[(l,r)]

        for i in range(len(s)):
            res += dp(i,i)
            if i <= len(s) - 2:
                res += dp(i, i+1)
        return res